chain/chain/src/store/latest_witnesses.rs

```diff
@@ -0,0 +1,149 @@
+//! This module is responsible for storing the latest ChunkStateWitnesses observed
+//! by the node. The latest witnesses are stored in the database and can be fetched
+//! for analysis and debugging.
+//! The number of stored witnesses is limited. When the limit is reached
+//! the witness with the oldest height is removed from the database.
+//! At the moment this module is used only for debugging purposes.
+
+use near_primitives::stateless_validation::ChunkStateWitness;
+use near_store::DBCol;
+
+use crate::ChainStoreAccess;
+
+use super::ChainStore;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_store::db::LATEST_WITNESSES_INFO;
+use rand::rngs::OsRng;
+use rand::RngCore;
+
+/// Maximum size of the latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_SIZE: ByteSize = ByteSize::gb(4);
+
+/// Maximum size of a single latest witness stored in the database.
+const SINGLE_LATEST_WITNESS_MAX_SIZE: ByteSize = ByteSize::mb(128);
+
+/// Maximum number of latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_COUNT: u64 = 60 * 30;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct LatestWitnessesKey {
+    pub height: u64,
+    pub shard_id: u64,
+    /// Each witness has a random UUID to ensure that the key is unique.
+    /// It allows to store multiple witnesses with the same height and shard_id.
+    pub random_uuid: [u8; 16],
```

Comment:
can this be the account id of the chunk producer that sent the witness? from chunk producer side it will be single, from chunk validator side it should expect a single witness from each chunk producer.

Reply:
We could, but I'm not sure if there's any benefit to it.
Usually there'll be only one witness per (height, shard_id). The situations where there is more than one will be weird, where anything can happen, including sending multiple witnesses by the same chunk producer.
For debugging I think it's best to save everything that we receive, without deduplication. If I were debugging witness issues I'd like to know that the chunk producer is sending duplicate messages, and deduplicating by account id would hide that from me.

Reply:
ok, yes, using account id would hide the case of the same chunk producer sending duplicate witnesses. and witness already has the account it for debugging.

Reply:
> I thought @pugachAG was working on https://github.com/near/nearcore/pull/11111 where we only processed one state witness for a given height, epoch, shard? This shouldn't then be needed?

I wanted to put saving witnesses before Anton's check, debug information should contain all witnesses, even the ones that were rejected.

Reply:
> We should include epoch_id in the key as well

I think it's impossible to have multiple possible epochs at a height. There's exactly one valid block per height, and this block will always have a clearly defined epoch.

Reply:
Ok, but we could receive witnesses on different epochs. I can add an epoch filter as well.